docs: replace machine charm tutorial by an improved tutorial (#2119)
canonical/operator: docs/tutorial/write-a-machine-charm-for-tinyproxy.md

Next, open a terminal, then run:

```text
multipass launch --cpus 4 --memory 8G --disk 50G --name juju-sandbox
```

This creates a virtual machine called `juju-sandbox`.

Multipass allocates some of your computer's memory and disk space to your virtual machine. The options we've chosen for `multipass launch` ensure that your virtual machine will be powerful enough to run Juju and deploy medium-sized charms.

This step should take less than 10 minutes, but the time depends on your computer and network. When your virtual machine has been created, you'll see the message:

```text
Launched: juju-sandbox
```

Although you'll test and deploy your charm inside your virtual machine, you'll probably find it more convenient to write your charm using your usual text editor or IDE. Let's use the Multipass {external+multipass:ref}`mount <reference-command-line-interface-mount>` command to make a directory available inside your virtual machine:

```text
multipass stop juju-sandbox
mkdir ~/tinyproxy-tutorial  # You'll write your charm in this directory.
multipass mount --type native ~/tinyproxy-tutorial juju-sandbox:~/tinyproxy

> ```

Reviewer: Per discussion with @dwilding elsewhere, we're thinking about changing the order of operations to:
1. create VM
2. shell
3. concierge and friends
4. stop
5. snapshot
6. mount

This lets us take a snapshot at a useful point, and removes the need to wrangle `PATH` for `uv install tox` (as @dwilding noticed) -- and we need to stop the VM before mounting anyway.

Author: This approach works, thanks for the suggestion! I've updated the [preview tutorial](https://canonical-ubuntu-documentation-library--2119.com.readthedocs.build/ops/2119/tutorial/write-a-machine-charm-for-tinyproxy/). Would you mind reading through from "To install tox" to the "Create a charm project" section?

I've addressed your other feedback too, BTW

Reviewer: The changes look great, thanks for addressing all my feedback.

## Write your charm

### Edit the metadata

Open `~/tinyproxy-tutorial/charmcraft.yaml` in your usual text editor or IDE, then change the values of `title`, `summary`, and `description` to:

```yaml
title: Reverse Proxy Demo
summary: A demo charm that configures tinyproxy as a reverse proxy.
description: |
  This charm demonstrates how to write a machine charm with Ops.
```

### Write a helper module

Your charm will interact with tinyproxy, so it's a good idea to write a helper module that wraps tinyproxy.
Charmcraft created `src/tinyproxy.py` as a placeholder helper module.

The helper module will be independent of the main logic of your charm. This will make it easier to test your charm. However, the helper module won't be a general-purpose wrapper for tinyproxy. The helper module will contain opinionated functions for managing tinyproxy on Ubuntu.

The helper module will depend on some libraries that are useful when writing charms.

To add the libraries to your charm's dependencies, run:

```text
uv add charmlibs-apt charmlibs-pathops
```

> This has added the following Python packages to the `dependencies` list in `pyproject.toml`:

Reviewer: ```suggestion
This adds the following Python packages to the `dependencies` list in `pyproject.toml`:
```
Not sure what tense is best here, maybe you have it right already.

Author: I dunno, I think what I have is not wrong, but when I read it I prefer your version

- {external+charmlibs:ref}`charmlibs-pathops <charmlibs-pathops>` - A file operations library, similar to `pathlib` from the standard library.

Next, replace the contents of `src/tinyproxy.py` with:

```python
"""Functions for interacting with tinyproxy."""

import logging
import os
import shutil
import signal
import subprocess

from charmlibs import apt, pathops

logger = logging.getLogger(__name__)

CONFIG_FILE = "/etc/tinyproxy/tinyproxy.conf"
PID_FILE = "/var/run/tinyproxy.pid"

def ensure_config(port: int, slug: str) -> bool:
    """Ensure that tinyproxy is configured. Return True if any changes were made."""
    # For the config file format, see https://manpages.ubuntu.com/manpages/jammy/en/man5/tinyproxy.conf.5.html
    config = f"""\
PidFile "{PID_FILE}"
Port {port}
Timeout 600
ReverseOnly Yes
ReversePath "/{slug}/" "http://www.example.com/"
"""
    return pathops.ensure_contents(CONFIG_FILE, config)

def get_version() -> str:
    """Get the version of tinyproxy that is installed."""
    result = subprocess.run(["tinyproxy", "-v"], check=True, capture_output=True, text=True)
    return result.stdout.removeprefix("tinyproxy").strip()

def install() -> None:

>     """Use APT to install the tinyproxy executable."""

Reviewer: Feel odd to see `apt` styled this way, but I guess it's correct --  actually I think this is my first time learning that it's an acronym.

```python
    def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
        """Handle config-changed event."""
        self.configure_and_run()
```

### Report status to Juju

Your charm should report the machine's status to Juju after handling each event. This enables you to use Juju's status output to see whether tinyproxy is running and correctly configured.

One option would be to modify each method in the charm class to report an appropriate status. However, this tends to be awkward in practice, because the logic to decide which status is most appropriate can become complex. Instead, we'll handle a special "collect-unit-status" event that is produced by Ops.

Add the following line to the `__init__` method of the charm class:

```python
        framework.observe(self.on.collect_unit_status, self._on_collect_status)
```

Then add the following method to the charm class:

```python
    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
        """Report the status of tinyproxy (runs after each event)."""
        try:
            self.load_config(TinyproxyConfig)
        except pydantic.ValidationError as e:

>             slug_error = e.errors()[0]  # Index 0 because 'slug' is the only option validated.

Reviewer: Very minor style thing, perhaps just my personal taste:
```suggestion
            slug_error, *_ = e.errors()  # 'slug' is the first and only option validated.
```
Or if this is correct:
```suggestion
            (slug_error,) = e.errors()  # 'slug' is the first and only option validated.
```

When Charmcraft has packed your charm, you'll see a message similar to:

```text
Packed tinyproxy_amd64.charm
```

The name of the `.charm` file depends on your computer's architecture. For example, if your computer has an ARM-based architecture, the file is called `tinyproxy_arm64.charm`.

### Deploy your charm

As you deploy your charm to Juju, it will be helpful to watch Juju status in real time.

Open another terminal, then run:

```text
multipass shell juju-sandbox
```

Next, in the same terminal, run:

```text
juju status --watch 2s
```

You should now have two terminals:

- A terminal with working directory `~/tinyproxy`, from earlier.

- A terminal that shows Juju status in real time:

    ```text
    Model    Controller     Cloud/Region         Version  SLA          Timestamp
    testing  concierge-lxd  localhost/localhost  3.6.8    unsupported  09:00:00+08:00
    ```

You're now ready to deploy your charm.

In the `~/tinyproxy` directory, run `juju deploy ./<charm-file>`, where `<charm-file`> is the name of the file created by `charmcraft pack`. For example:

```text
juju deploy ./tinyproxy_amd64.charm
```

> Juju creates an "application" from your charm. For each unit in the application, Juju starts a machine and installs your charm on the machine. We didn't tell Juju how many units we want, so Juju assumes one unit and starts one machine. After Juju has installed your charm on the machine, Juju starts sending events to your charm so that your charm can install and start tinyproxy.

Reviewer: ```suggestion
Juju creates an "application" from your charm. For each unit in the application, Juju starts a LXD virtual machine and installs your charm on the machine. We didn't tell Juju how many units we want, so Juju assumes one unit and starts one machine. After Juju has installed your charm on the machine, Juju starts sending events to your charm so that your charm can install and start tinyproxy.
```

TOTAL                119     31     26      7    70%
  unit: OK (1.21=setup[0.05]+cmd[1.03,0.13] seconds)
  congratulations :) (1.30 seconds)
```

## Write integration tests

Integration tests are an important way to check that your charm works correctly when deployed. In contrast to unit tests, integration tests require Juju to be available, and events aren't simulated.

When you created the initial version of your charm, Charmcraft included integration tests. The tests use {external+jubilant:doc}`Jubilant <index>` to interact with Juju. We'll expand the tests to cover more of your charm's functionality.

In `tests/integration/test_charm.py`, change `juju.wait(jubilant.all_active)` to:

```python
juju.wait(jubilant.all_active, timeout=600)
```

This extends the duration that Jubilant waits for your charm to deploy, in case the integration tests run slowly in your virtual machine. The default duration would be sufficient if the integration tests were running in a continuous integration environment.

Next, remove the `@pytest.mark.skip` decorator from `test_workload_version_is_set`. Then change `assert version == ...` to:

```python

>     assert version == "1.11.0"

Reviewer: ```suggestion
    assert version == "1.11.0"  # the version installed by tinyproxy.install
```

If you wanted isolated tests, you could change `juju` to be function-scoped (pytest's default scope) and deploy the `.charm` file at the beginning of each test. However, this would slow down the tests.

Now run the tests:

```text
tox -e integration
```

It will take a few minutes to run the tests. The output should be similar to:

```text
...

======================= 3 passed in 277.23s (0:04:37) =======================
  integration: OK (277.76=setup[0.06]+cmd[277.70] seconds)
  congratulations :) (277.89 seconds)
```

```{tip}
`tox -e integration` doesn't pack your charm. If you modify the charm code and want to run the integration tests again, run `charmcraft pack` before `tox -e integration`.
```

## Tear things down

Congratulations on reaching the end of the tutorial!

You can keep things running, to explore further, or you can remove what you created:

- If you're still watching Juju status, press <kbd>Ctrl</kbd> + <kbd>C</kbd> to stop watching.

> - To remove your charm from Juju, run `juju remove-application tinyproxy`.

Reviewer: I wonder if it's worth mentioning that it doesn't matter if these steps go wrong if you're going to delete the VM anyway?

Author: I've added a sentence to say that you don't need to remove the charm if you're going to remove the VM.

I don't think it's worth explaining how to unsetup the VM - out of scope, I'd say.

- **uv and tox** - You'll implement your charm using Python code. [uv](https://docs.astral.sh/uv/) is a Python project manager that will install dependencies for checks and tests. You'll use [tox](https://tox.wiki/en/) to select which checks or tests to run.

Instead of manually installing and configuring each tool, we recommend using [Concierge](https://github.com/canonical/concierge), Canonical's tool for setting up charm development environments.

In your virtual machine, run:

```text
sudo snap install --classic concierge
sudo concierge prepare -p machine --extra-snaps astral-uv --charmcraft-channel latest/candidate
```

This first installs Concierge, then uses Concierge to install and configure the other tools (except tox). The option `-p machine` tells Concierge that we want tools for developing machine charms.

It will take up to 15 minutes to install the tools, depending on your computer and network. When the tools have been installed, you'll see a message that ends with:

```text
msg="Bootstrapped Juju" provider=lxd
```

To install tox, run:

```text
uv tool install tox --with tox-uv
export PATH="$HOME/.local/bin:$PATH"
```

> 

Reviewer: `````suggestion

````{tip}
Take a snapshot of your VM now so that you can easily restart the tutorial steps from scratch without waiting for the tools to be reinstalled.
```text
juju-sandbox juju-sandbox --name=juju-sandbox-with-tools
```
````

`````

Your charm will interact with tinyproxy, so it's a good idea to write a helper module that wraps tinyproxy.
Charmcraft created `src/tinyproxy.py` as a placeholder helper module.

The helper module will be independent of the main logic of your charm. This will make it easier to test your charm. However, the helper module won't be a general-purpose wrapper for tinyproxy. The helper module will contain opinionated functions for managing tinyproxy on Ubuntu.

The helper module will depend on some libraries that are useful when writing charms.

To add the libraries to your charm's dependencies, run:

```text
uv add charmlibs-apt charmlibs-pathops
```

This has added the following Python packages to the `dependencies` list in `pyproject.toml`:

- [`charmlibs-apt`](https://documentation.ubuntu.com/charmlibs/reference/charmlibs/apt/) - A library for using APT to manage system packages. This is how your charm will install tinyproxy.
- [`charmlibs-pathops`](https://documentation.ubuntu.com/charmlibs/reference/charmlibs/pathops/) - A file operations library, similar to `pathlib` from the standard library.

Next, replace the contents of `src/tinyproxy.py` with:

```python

> import logging

Author: Resolved, thanks!

Unit          Workload  Agent  Machine  Public address  Ports  Message
tinyproxy/0*  active    idle   0        10.71.67.208

Machine  State    Address       Inst id        Base          AZ  Message
0        started  10.71.67.208  juju-8e7bd9-0  ubuntu@22.04      Running
```

```{tip}
For the rest of the tutorial, we'll assume that you're still watching Juju status. To stop watching, press <kbd>Ctrl</kbd> + <kbd>C</kbd>.
```

### Try the reverse proxy

Now that tinyproxy is running, we can check that it proxies [example.com](http://example.com) on the machine's network.

In your virtual machine, run:

```text
curl <address>:8000/example/
```

Where `<address>` is the IP address of machine 0 from Juju status. In our example of Juju status, the IP address is 10.71.67.208.

The output of curl should be similar to:

```text
<!doctype html><html lang="en"><head><title>Example Domain</title>...
```

Then run `curl http://example.com` and check that you get the same output.

### Change the configuration

Let's see what happens to the reverse proxy if we change the `slug` configuration option. Run:

```text
juju config tinyproxy slug=foo
```

> Where `<address>` is the IP address of machine 0 from Juju status.

Reviewer 2: This doesn't seem to make sense here (just after the `juju config` command). Probably don't need it again after the above.

When you created the initial version of your charm, Charmcraft included integration tests. The tests use {external+jubilant:doc}`Jubilant <index>` to interact with Juju. We'll expand the tests to cover more of your charm's functionality.

In `tests/integration/test_charm.py`, change `juju.wait(jubilant.all_active)` to:

```python
juju.wait(jubilant.all_active, timeout=600)
```

This extends the duration that Jubilant waits for your charm to deploy, in case the integration tests run slowly in your virtual machine. The default duration would be sufficient if the integration tests were running in a continuous integration environment.

Next, remove the `@pytest.mark.skip` decorator from `test_workload_version_is_set`. Then change `assert version == ...` to:

```python
    assert version == "1.11.0"
```

You should now have the following tests:

- `test_deploy` - Deploys your charm and checks that it goes into active status.
- `test_workload_version_is_set` - Checks that your charm reports the correct version of tinyproxy to Juju.

> You might have a question at this point: what makes 1.11.0 the "correct" version of tinyproxy? The answer is, we're cheating slightly because we happen to know that APT will install version 1.11.0 on the machine. In general, it's good practice for charms to pin workload versions, so the charm code ought to ask APT to install a specific version of tinyproxy.

Author: Thanks for raising this. I agree it's better to show the best practice.

I've made a few changes:

- Updated the function that installs tinyproxy:

    ```py
    # Install a specific package from ubuntu@22.04
    # See https://packages.ubuntu.com/jammy/tinyproxy-bin
    # In general, it's good practice for charms to pin workload versions.
    apt.add_package("tinyproxy-bin", "1.11.0-1")
    ```

- Removed the paragraph that talks about "cheating"

- In the unit tests that use a mock version number, deliberately used "1.0.0" instead of the real version number, so that it's clearer where the pinned version number needs to go.

Then save the file and exit nano.

````{tip}
You'll need to edit Python files later in the tutorial. To configure nano to insert spaces instead of tabs, run:

```text
printf "set tabsize 4\nset tabstospaces\n" > ~/.nanorc
```
````

### Write a helper module

Your charm will interact with tinyproxy, so it's a good idea to write a helper module that wraps tinyproxy.
Charmcraft created `src/tinyproxy.py` as a placeholder helper module.

The helper module will be independent of the main logic of your charm. This will make it easier to test your charm. However, the helper module won't be a general-purpose wrapper for tinyproxy. The helper module will contain opinionated functions for managing tinyproxy on Ubuntu.

The helper module will depend on some libraries that are useful when writing charms.

To add the libraries to your charm's dependencies, run:

```text
uv add charmlibs-apt charmlibs-pathops
```

This has added the following Python packages to the `dependencies` list in `pyproject.toml`:

- [`charmlibs-apt`](https://documentation.ubuntu.com/charmlibs/reference/charmlibs/apt/) - A library for using APT to manage system packages. This is how your charm will install tinyproxy.

> - [`charmlibs-pathops`](https://documentation.ubuntu.com/charmlibs/reference/charmlibs/pathops/) - A file operations library, similar to `pathlib` from the standard library.

Author: @james-garner-canonical do you know what the intersphinx targets would be for these? I had a quick try, but couldn't figure out the right targets.

Author: James is adding targets in https://github.com/canonical/charmlibs/pull/233. I'll switch to intersphinx links after that PR merges

Instead of manually installing and configuring each tool, we recommend using [Concierge](https://github.com/canonical/concierge), Canonical's tool for setting up charm development environments.

In your virtual machine, run:

```text
sudo snap install --classic concierge
sudo concierge prepare -p machine --extra-snaps astral-uv --charmcraft-channel latest/candidate
```

This first installs Concierge, then uses Concierge to install and configure the other tools (except tox). The option `-p machine` tells Concierge that we want tools for developing machine charms.

It will take up to 15 minutes to install the tools, depending on your computer and network. When the tools have been installed, you'll see a message that ends with:

```text
msg="Bootstrapped Juju" provider=lxd
```

To install tox, run:

```text
uv tool install tox --with tox-uv
uv tool update-shell
```

When tox has been installed, you'll see the message:

```text
Restart your shell to apply changes
```

It's important to exit your virtual machine at this point. If you don't exit your virtual machine, you won't be able to use tox later in the tutorial.

To exit your virtual machine, run:

```text
exit
```

> The terminal switches back to your usual operating system. Your virtual machine is still running in the background.

Reviewer 3: > Does `uv tool update-shell` not eliminate the need for this?

No, I wondered about that too. `update-shell` updates things so that when you re-start the shell everything is available. You could do it in other ways, like `exec bash` or `source (the file that changed, probably bashrc?)` or something, but I guess this is probably simplest even if not most convenient.

## Create a charm project

You'll need to work inside your virtual machine again. Open a terminal, then run:

```text
multipass shell juju-sandbox
```

The terminal prompt is now `ubuntu@juju-sandbox:~$`.

Next, create a project directory and use Charmcraft to create the initial version of your charm:

```text
mkdir tinyproxy
cd tinyproxy
charmcraft init --profile machine

> ```

Reviewer: Are we sure we want to do this inside the VM? I recall an older tutorial (the previous machine one?) had you create files locally and mount the directory in the VM. This was nice as it let you trivially blow the VM away if it got into a bad state without losing any of your work following the tutorial.

def stop() -> None:
    """Stop tinyproxy."""
    pid = _get_pid()
    if pid:
        os.kill(pid, signal.SIGTERM)

def uninstall() -> None:
    """Uninstall the tinyproxy executable and remove files."""
    apt.remove_package("tinyproxy-bin")
    pathops.LocalPath(PID_FILE).unlink(missing_ok=True)
    pathops.LocalPath(CONFIG_FILE).unlink(missing_ok=True)
    pathops.LocalPath(CONFIG_FILE).parent.rmdir()

def _get_pid() -> int | None:
    """Return the PID of the tinyproxy process, or None if the process can't be found."""
    if not pathops.LocalPath(PID_FILE).exists():
        return None
    pid = int(pathops.LocalPath(PID_FILE).read_text())
    try:
        # Sending signal 0 doesn't terminate the process. It just checks whether the PID exists.
        os.kill(pid, 0)
    except ProcessLookupError:
        return None
    return pid
```

Notice that the helper module is stateless. In fact, your charm as a whole will be stateless. The main logic of your charm will:

1. Receive an event from Juju.
2. Use the functions in the helper module to manage tinyproxy and check its status.
3. Report the status back to Juju.

```{tip}

> After adding code to your charm, run `tox -e format` to format the code. Then run `tox -e lint` to check the code against coding style standards and run static checks.

Reviewer 3: It's a bit unfortunate that if you run `tox -e format` at this point you get an error, because there's no docstring. I know we want to keep the code snippets small, but maybe it would be worth having a one-liner version just so that the format check exits without errors?

Author: Ah, thanks for catching this. The full module in the code has a docstring, but I'd missed that I need to show that in the tutorial too. Updated.

```

### Handle the install event

We'll now write the charm code that handles events from Juju. Charmcraft created `src/charm.py` as the location for this logic, containing a class called `TinyproxyCharm`. We'll refer to `TinyproxyCharm` as the "charm class".

In `src/charm.py`, replace the `_on_install` method of the charm class with:

```python
    def _on_install(self, event: ops.InstallEvent) -> None:
        """Install tinyproxy on the machine."""
        if not tinyproxy.is_installed():
            tinyproxy.install()
            version = tinyproxy.get_version()
            self.unit.set_workload_version(version)
```

> When your charm receives the "install" event from Juju, Ops runs this method and responds to Juju with the version of tinyproxy that's installed on the machine. Juju shows the tinyproxy version in its status output.

Reviewer 3: "responds to Juju" doesn't seem right to me - it's not the "response" of the install event. What about something like "and tells Juju the version", or "responds by telling Juju the version".

To handle these events, add the following lines to the `__init__` method of the charm class:

```python
        framework.observe(self.on.stop, self._on_stop)
        framework.observe(self.on.remove, self._on_remove)
```

Then add the following methods to the charm class:

```python
    def _on_stop(self, event: ops.StopEvent) -> None:
        """Handle stop event."""
        tinyproxy.stop()
        self.wait_for_not_running()

    def _on_remove(self, event: ops.RemoveEvent) -> None:
        """Handle remove event."""
        tinyproxy.uninstall()

    def wait_for_not_running(self) -> None:
        """Wait for tinyproxy to not be running."""
        for _ in range(3):
            if not tinyproxy.is_running():
                return
            time.sleep(1)
        raise RuntimeError("tinyproxy was still running after the expected time")
```

## Try your charm

### Pack your charm

Before you can try your charm, you need to "pack" it. Packing combines the charm code and metadata into a single file that can be deployed to Juju.

In your virtual machine, make sure that the working directory is `~/tinyproxy`. Then run:

```text
charmcraft pack
```

> Charmcraft will take about 20 minutes to pack your charm, depending on your computer and network.

Reviewer 3: Since we're using the `uv` plugin and exploiting the bug to avoid building `pydantic`, this seems excessive. I do find that it's regularly about 20 minutes if you're building everything from source, but this took 8m35s for me because we skipped that. Maybe we should be less certain about this and say it could take as much as 20 minutes, or something like that?

Reviewer 2: Yeah, this took 2m29s for me.

        self.installed = True

    def is_installed(self) -> bool:
        return self.installed

    def is_running(self) -> bool:
        return self.running

    def start(self) -> None:
        self.running = True

    def stop(self) -> None:
        self.running = False

    def reload_config(self) -> None:
        self.reloaded_config = True

    def uninstall(self) -> None:
        self.installed = False

def patch_charm(monkeypatch: pytest.MonkeyPatch, tinyproxy: MockTinyproxy):
    """Patch the helper module to use mock functions for interacting with tinyproxy."""
    monkeypatch.setattr("charm.tinyproxy.ensure_config", tinyproxy.ensure_config)
    monkeypatch.setattr("charm.tinyproxy.get_version", tinyproxy.get_version)
    monkeypatch.setattr("charm.tinyproxy.install", tinyproxy.install)
    monkeypatch.setattr("charm.tinyproxy.is_installed", tinyproxy.is_installed)
    monkeypatch.setattr("charm.tinyproxy.is_running", tinyproxy.is_running)
    monkeypatch.setattr("charm.tinyproxy.start", tinyproxy.start)
    monkeypatch.setattr("charm.tinyproxy.stop", tinyproxy.stop)
    monkeypatch.setattr("charm.tinyproxy.reload_config", tinyproxy.reload_config)

>     monkeypatch.setattr("charm.tinyproxy.uninstall", tinyproxy.uninstall)

Reviewer 3: I didn't test this, so maybe it doesn't work, but could you just monkeypatch `charm.tinyproxy` to be `MockTinyproxy` here instead of individually doing each method?

Author: Wow, this makes things much nicer! Why didn't I think of that :laughing: 

I've been able to replace the whole `patch_charm` function by:

```py
monkeypatch.setattr("charm.tinyproxy", tinyproxy)
```
